File: benchmarks/pandas/bench_attrs_advanced.py
"""Benchmark: pandas Series attrs advanced — individual attr get/set/delete/copy/merge"""
import json, time
import pandas as pd

WARMUP = 3
ITERATIONS = 1_000

s = pd.Series(range(1_000))
s2 = pd.Series(range(1_000))

for _ in range(WARMUP):
    s.attrs["unit"] = "meters"
    _ = s.attrs.get("unit")
    _ = bool(s.attrs)
    s2.attrs.update(dict(s.attrs))
    s.attrs.update({"version": 1})
    s.attrs.pop("unit", None)
    s.attrs.clear()

start = time.perf_counter()
for i in range(ITERATIONS):
    s.attrs["unit"] = "meters"
    _ = s.attrs.get("unit")
    _ = bool(s.attrs)
    s2.attrs.update(dict(s.attrs))
    s.attrs.update({"version": i})
    s.attrs.pop("unit", None)
    s.attrs.clear()
total = (time.perf_counter() - start) * 1000
print(json.dumps({"function": "attrs_advanced", "mean_ms": total / ITERATIONS, "iterations": ITERATIONS, "total_ms": total}))

File: benchmarks/tsb/bench_attrs_advanced.ts
/**
 * Benchmark: advanced attrs helpers — getAttr/setAttr/deleteAttr/clearAttrs/copyAttrs/mergeAttrs/hasAttrs
 */
import { Series, getAttr, setAttr, deleteAttr, clearAttrs, copyAttrs, mergeAttrs, hasAttrs } from "../../src/index.js";

const N = 1_000;
const s = new Series({ data: Array.from({ length: N }, (_, i) => i) });
const s2 = new Series({ data: Array.from({ length: N }, (_, i) => i * 2) });

const WARMUP = 3;
const ITERATIONS = 1_000;

for (let i = 0; i < WARMUP; i++) {
  setAttr(s, "unit", "meters");
  getAttr(s, "unit");
  hasAttrs(s);
  copyAttrs(s, s2);
  mergeAttrs(s, { version: 1 });
  deleteAttr(s, "unit");
  clearAttrs(s);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  setAttr(s, "unit", "meters");
  getAttr(s, "unit");
  hasAttrs(s);
  copyAttrs(s, s2);
  mergeAttrs(s, { version: i });
  deleteAttr(s, "unit");
  clearAttrs(s);
}
const total = performance.now() - start;
console.log(
  JSON.stringify({
    function: "attrs_advanced",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
